pkg/apis/storage/types.go

```diff
@@ -363,11 +377,14 @@ type CSIDriverSpec struct {
 	// most one token is empty string. To receive a new token after expiry,
 	// RequiresRepublish can be used to trigger NodePublishVolume periodically.
 	//
-	// This is an alpha feature and only available when the
+	// This is a beta feature and only available when the
 	// CSIServiceAccountToken feature is enabled.
 	//
 	// +optional
-	// +listType=atomic
+	// +patchMergeKey=audience
```

Comment:
I don't think a mergeable list is what we want here (and it actually doesn't work consistently in my experiments).

The audience field is not actually required by the API server (you can `kubectl apply -f ... --validate=false` a manifest that is missing it, or create one via a controller without explicitly sending an audience field in a tokenrequest item)

If you do that, you end up with an existing persisted object `kubectl apply` and `kubectl apply --server-side` don't know how to merge updates to properly.

They either complain and fail:
```
kubectl apply -f ~/snippets/csidriver/driver2.yaml
warning: error calculating patch from openapi spec: map: map[] does not contain declared merge key: audience
error: error when applying patch:

to:
Resource: "storage.k8s.io/v1, Resource=csidrivers", GroupVersionKind: "storage.k8s.io/v1, Kind=CSIDriver"
Name: "tokens", Namespace: ""
for: "/Users/liggitt/snippets/csidriver/driver2.yaml": error when creating patch with:
original:
{"apiVersion":"storage.k8s.io/v1","kind":"CSIDriver","metadata":{"annotations":{},"name":"tokens"},"spec":{"tokenRequests":[{}],"volumeLifecycleModes":["Persistent","Ephemeral"]}}

modified:
{"apiVersion":"storage.k8s.io/v1","kind":"CSIDriver","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"storage.k8s.io/v1\",\"kind\":\"CSIDriver\",\"metadata\":{\"annotations\":{},\"name\":\"tokens\"},\"spec\":{\"tokenRequests\":[{\"audience\":\"another\"}],\"volumeLifecycleModes\":[\"Persistent\",\"Ephemeral\"]}}\n"},"name":"tokens"},"spec":{"tokenRequests":[{"audience":"another"}],"volumeLifecycleModes":["Persistent","Ephemeral"]}}

current:
{"apiVersion":"storage.k8s.io/v1","kind":"CSIDriver","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"storage.k8s.io/v1\",\"kind\":\"CSIDriver\",\"metadata\":{\"annotations\":{},\"name\":\"tokens\"},\"spec\":{\"tokenRequests\":[{}],\"volumeLifecycleModes\":[\"Persistent\",\"Ephemeral\"]}}\n"},"creationTimestamp":"2021-03-09T20:57:24Z","managedFields":[{"apiVersion":"storage.k8s.io/v1","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:annotations":{".":{},"f:kubectl.kubernetes.io/last-applied-configuration":{}}},"f:spec":{"f:attachRequired":{},"f:fsGroupPolicy":{},"f:podInfoOnMount":{},"f:requiresRepublish":{},"f:storageCapacity":{},"f:tokenRequests":{".":{},"k:{\"audience\":\"\"}":{".":{},"f:audience":{}}},"f:volumeLifecycleModes":{".":{},"v:\"Ephemeral\"":{},"v:\"Persistent\"":{}}}},"manager":"kubectl-client-side-apply","operation":"Update","time":"2021-03-09T20:57:24Z"}],"name":"tokens","resourceVersion":"330","uid":"98c2d9e1-817f-40a0-b2ba-fae511f4adb1"},"spec":{"attachRequired":true,"fsGroupPolicy":"ReadWriteOnceWithFSType","podInfoOnMount":false,"requiresRepublish":false,"storageCapacity":false,"tokenRequests":[{"audience":""}],"volumeLifecycleModes":["Persistent","Ephemeral"]}}

for: "/Users/liggitt/snippets/csidriver/driver2.yaml": map: map[] does not contain declared merge key: audience
```

or silently keep the existing `audience:""` entry and add another item to the list.

I would expect a single actor to be defining the CSIDriver spec and dictating the entire content of this list, which would mean it should stay atomic.

Reply:
`kubectl apply --validate=false` with this:
```
apiVersion: storage.k8s.io/v1
kind: CSIDriver
metadata:
  name: tokens
spec:
  volumeLifecycleModes:
  - Persistent
  - Ephemeral
  tokenRequests: [{}]
```

then try to `kubectl apply` with this:
```
apiVersion: storage.k8s.io/v1
kind: CSIDriver
metadata:
  name: tokens
spec:
  volumeLifecycleModes:
  - Persistent
  - Ephemeral
  tokenRequests: [{"audience":"another"}]
```

`kubectl apply --server-side` submits the request, but produces a CSIDriver containing the old audience as well, rather than dropping it properly:
```yaml
  tokenRequests:
  - audience: another
  - audience: ""
```

Reply:
didn't know there is a `--validate=false` to ignore the validation. how to mark a field as required if`--validate=false`? then i think above case would be prevented.

i was thinking there might be a case where a controller is reconciling the CSIDriver object but yea, single actor is most common. 

changed back to atomic.